File: client-android/lib/src/main/kotlin/stasis/client_android/lib/tracking/RecoveryTracker.kt
package stasis.client_android.lib.tracking

import stasis.client_android.lib.model.EntityRef
import stasis.client_android.lib.model.TargetEntity
import stasis.client_android.lib.ops.OperationId
import stasis.client_android.lib.tracking.state.RecoveryState

interface RecoveryTracker {
    fun started(operation: OperationId)
    fun entityExamined(operation: OperationId, entity: EntityRef, metadataChanged: Boolean, contentChanged: Boolean)
    fun entityCollected(operation: OperationId, entity: TargetEntity)
    fun entityProcessingStarted(operation: OperationId, entity: EntityRef, expectedParts: Int)
    fun entityPartProcessed(operation: OperationId, entity: EntityRef)
    fun entityProcessed(operation: OperationId, entity: EntityRef)
    fun metadataApplied(operation: OperationId, entity: EntityRef)
    fun failureEncountered(operation: OperationId, failure: Throwable)
    fun failureEncountered(operation: OperationId, entity: EntityRef, failure: Throwable)
    fun completed(operation: OperationId)

    suspend fun stateOf(operation: OperationId): RecoveryState?
}

File: client-ios/lib/Sources/StasisClientLib/Tracking/RecoveryTracker.swift
import Foundation

public protocol RecoveryTracker: Sendable {
    func started(operation: OperationId) async
    func entityExamined(operation: OperationId, entity: URL, metadataChanged: Bool, contentChanged: Bool) async
    func entityCollected(operation: OperationId, entity: TargetEntity) async
    func entityProcessingStarted(operation: OperationId, entity: URL, expectedParts: Int) async
    func entityPartProcessed(operation: OperationId, entity: URL) async
    func entityProcessed(operation: OperationId, entity: URL) async
    func metadataApplied(operation: OperationId, entity: URL) async
    func failureEncountered(operation: OperationId, failure: Error) async
    func failureEncountered(operation: OperationId, entity: URL, failure: Error) async
    func completed(operation: OperationId) async

    func stateOf(operation: OperationId) async -> RecoveryState?
}
